### PR TITLE
feat: implement sas file system for viya connections

### PR DESCRIPTION
**Pull request by scottdover**
**Summary**
This adds sas file system support for viya connections. Notable changes include:
 - Introducing a `canRecycleResource` function to sas content adapter. This allows us to make a determination about whether or not we should show a dialog for deleted files. Since SAS file system doesn't support recycle bin, we show the deletion message every time a file is deleted (since it's a permanent deletion)
 - Creates a distinct connection for `NotebookConverter` for connecting to sas studio (instead of re-using content model's connection)
 - Favorites are not implemented in this pull request and will be implemented in a future PR

**Testing**
 - [x] File/folder creation
    - [x] Create file/folder w/ context menu
    - [x] Create file/folder by upload
    - [x] Create file/folder by drag & drop (create multiple files)
 - [x] File/folder deletion
    - [x] Test file deletion with context menu
    - [x] Test multi-file deletion with context menu
 - [x] File/folder updates
    - [x] Test updating file/folder name
    - [x] Test updating file contents
    - [x] Test moving file/folder (multiple files/folders)
 - [x] Test downloading files/folders
 - [x] Make sure refresh works as expected
    - [x] Make sure connections are automatically refreshed after they become stale
 - [x] Make sure a sas notebook file can be converted to a flow (test with sas content as well)
 - [x] Make sure we're displaying all files/folders for sas server and that items are sorted by type (directories before files), then alphabetically
 - [x] Make sure we can collapse all folders

**TODOs**
 - [x] Update CHANGELOG
 - [x] Update `matrix.md` with details about sas server

### CODE REVIEW

**Review by scnwwu — APPROVED** [review]
Looks good to me

**Comment by scottdover** [comment]
Just had to rebase for DCO things

**Comment by Zhirong2022** [comment]
Open VSC and the default connection profile is non-Viya initially,  switch to a Viya connection and sign in, it will show SAS content instead of SAS Sever stuff.
image

**Comment by Zhirong2022** [comment]
Error is thrown if trying to click Collapse All
image

Error running command SAS.server.collapseAllContent: Cannot read properties of undefined (reading 'treeIdentifier'). This is likely caused by the extension that contributes SAS.server.collapseAllContent.

**Comment by scnwwu** [comment]
@scottdover, you may want to update doc as appropriate, for example
https://github.com/sassoftware/vscode-sas-extension/blob/main/website/docs/matrix.md

**Comment by scottdover** [comment]
> @scottdover, you may want to update doc as appropriate, for example https://github.com/sassoftware/vscode-sas-extension/blob/main/website/docs/matrix.md

Ah, I knew I was forgetting something. Will update :)

**Comment by Zhirong2022** [comment]
Better error validation for some scenarios:
1.Create a folder with the same name as an existed one under the same directory
Current error message: Error running command SAS.server.addFolderResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.addFolderResource.

2.Folder name having some illegal special character like '/'
Current error message: Error running command SAS.server.addFolderResource: Request failed with status code 404. This is likely caused by the extension that contributes SAS.server.addFolderResource.

**Comment by Zhirong2022** [comment]
Create a **.txt** file named as  NewFile+!@$%^&*.txt, it failed to open it in VSC.  However it can be opened successfully in SASStudio.

**Comment by scottdover** [comment]
Hey @Zhirong2022 . All comments above should now be addressed.

@scnwwu, I've updated the changelog and documentation

File: website/docs/Features/accessServer.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+---
+sidebar_position: 1
```
Comment (scnwwu) [inline website/docs/Features/accessServer.md:2]: This affects the order of the sidebar entries and should not be duplicated.
accessContent.md is 1.
Would you please re-arrange the order as appropriate? You may have to change other files in this folder.

**Comment by Zhirong2022** [comment]
It is better to have a consistent file folder name violation with SASStudio. It is successfully to create a file folder with name 'NewFolder/' and it will be shown as 'NewFolder'.  
image

**Comment by Zhirong2022** [comment]
Multi-selection files or folders through Shift or Ctrl. Suppose to delete all the highlighted files or folders, it will ask the user to confirm the deletion one by one.

**Comment by Zhirong2022** [comment]
Create a file with the name
Error Message:
Error running command SAS.server.addFileResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.addFileResource.

Convert a .sasnb file to .flw file with the same name
1.SAS Server 
Error message:Request failed with status code 409
2.SAS Content
It will add _CopyX(sequence number) automatically to avoid having errors

**Comment by Zhirong2022** [comment]
Create a file named  as 'name#.sas' in SASStudio. It failed to open it in VSC with error 'he file is not displayed in the text editor because it is a directory.'

File: website/docs/Features/accessServer.md (hunk 1 above)
Comment (scottdover) [reply website/docs/Features/accessServer.md]: Ah, good catch. I think I got the order right in https://github.com/sassoftware/vscode-sas-extension/pull/1203/commits/37db855d5dc4b685119dc99727695530febed8eb

**Comment by scottdover** [comment]
> test#.sas

Hey @Zhirong2022. Let's skip this one in this pull request. This is currently an issue that impacts sas content & sas server. It looks to be an issue with uri parsing. I'll create a follow-up PR to handle this case.

**Comment by scottdover** [comment]
> Multi-selection files or folders through Shift or Ctrl. Suppose to delete all the highlighted files or folders, it will ask the user to confirm the deletion one by one.

This is intentional in this case. We warn the user about every deletion in SAS server because there is no recycle bin and deletion is permanent. You'll notice we do the same thing if you multi-select and delete from the recycle bin in sas content.

NOTE: This could certainly be improved to be a single message. That'll need to be handled in a separate PR though

**Comment by scottdover** [comment]
> Create a file with the same name Error Message: Error running command SAS.server.addFileResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.addFileResource.
> 
> Convert a .sasnb file to .flw file with the same name 1.SAS Server Error message:Request failed with status code 409 2.SAS Content It will add _CopyX(sequence number) automatically to avoid having errors

Hey @Zhirong2022 . This should now be resolved and ready to re-test

**Comment by Zhirong2022** [comment]
@scottdover , it failed to generate .vsix package, I got build error as bellow.
image

File: website/docs/Features/sasCodeEditing.md [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ---
-sidebar_position: 3
+sidebar_position: 6
```
Comment (scnwwu) [inline website/docs/Features/sasCodeEditing.md:2]: well, errorsWarnings.md is 6

File: client/src/components/ContentNavigator/ContentModel.ts [hunk 3]
```diff
@@ -90,6 +92,43 @@ export class ContentModel {
...
+      }, {});
+
+      if (!usedFlowNames[fileName]) {
+        return fileName;
+      }
+
+      let number = 1;
+      let newFileName;
+      do {
+        newFileName = l10n.t("{basename}_Copy{number}.flw", {
```
Comment (scnwwu) [inline client/src/components/ContentNavigator/ContentModel.ts:122]: nit: the function name looks general, use `${ext}` over `flw` may be better

**Comment by scottdover** [comment]
> @scottdover , it failed to generate .vsix package, I got build error as below. image

Sorry about that. Should be good to go now :)

**Comment by Zhirong2022** [comment]
Better error validation for scenarios below:
1.Rename a file with the same name as an existed one 
Error message: Error running command SAS.server.renameResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.renameResource.

2.Rename a file folder with the same name as an existed one
Error message:Error running command SAS.server.renameResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.renameResource.

**Comment by Zhirong2022** [comment]
It failed to save a file to the sever 
1.Create a file folder named **NewFolder**
2.Create a file named **NewFile.sas** under **NewFolder**
3.Type some code in the **NewFile.sas** file
4.Delete **NewFolder** and **NewFild.sas** was kept unsaved
5.Click SaveAs.. on the prompt message
6.Type the correct path which is authorized
image

**Comment by Zhirong2022** [comment]
Error is thrown if trying to convert the .sasnb file for the second time
1.Select a .sasnb file and convert it to a .flw file with the default given name, it works well
2.Switch the profile to a new one (any connection type)
3.Switch the profile back 
2.Delete the newly created .flw file 
3.Convert the same .sasnb the second time without change the .flw name

**Comment by Zhirong2022** [comment]
Error 'The file type is unsupported.'is thrown if trying to convert a .sasnb file 
1.Launch VSC and make sure the initial active profile is a non-Viya connection
2.Switch the profile to a Viya connection 
3.Create a new .sasnb file on file server
4.Convert the .sasnb file to the .flw file

**Comment by scottdover** [comment]
> Better error validation for scenarios below: 1.Rename a file with the same name as an existed one Error message: Error running command SAS.server.renameResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.renameResource.
> 
> 2.Rename a file folder with the same name as an existed one Error message:Error running command SAS.server.renameResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.renameResource.

- [ ] Resolved

> It failed to save a file to the sever 1.Create a file folder named **NewFolder** 2.Create a file named **NewFile.sas** under **NewFolder** 3.Type some code in the **NewFile.sas** file 4.Delete **NewFolder** and **NewFild.sas** was kept unsaved 5.Click SaveAs.. on the prompt message 6.Type the correct path which is authorized image

- [ ] Resolved

> Error is thrown if trying to convert the .sasnb file for the second time 1.Select a .sasnb file and convert it to a .flw file with the default given name, it works well 2.Switch the profile to a new one (any connection type) 3.Switch the profile back 4.Delete the newly created .flw file 5.Convert the same .sasnb the second time without change the .flw name

- [ ] Resolved

> Error 'The file type is unsupported.'is thrown if trying to convert a .sasnb file 1.Launch VSC and make sure the initial active profile is a non-Viya connection 2.Switch the profile to a Viya connection 3.Create a new .sasnb file on file server 4.Convert the .sasnb file to the .flw file

- [ ] Resolved

**Comment by Zhirong2022** [comment]
Create a new **.txt** file with name **!@$%^&*(**.txt, it has error if trying to download it through ContextMenu.
Error message:
Error running command SAS.server.downloadResource: Error: ENOENT: no such file or directory, open 'c:\\!@$%^&*(.txt'. This is likely caused by the extension that contributes SAS.server.downloadResource.

**Comment by scottdover** [comment]
Hey @Zhirong2022 . Things should be ready to test again now. See https://github.com/sassoftware/vscode-sas-extension/pull/1203#issuecomment-2452310484 for responses :)

**Comment by Zhirong2022** [comment]
@scottdover , please check the errors below with package generation.
image

**Comment by Zhirong2022** [comment]
> Create a new **.txt** file with name **!@$%^&*(**.txt, it has error if trying to download it through ContextMenu. Error message: Error running command SAS.server.downloadResource: Error: ENOENT: no such file or directory, open 'c:\!@$%^&*(.txt'. This is likely caused by the extension that contributes SAS.server.downloadResource.

@scottdover , I raised https://github.com/sassoftware/vscode-sas-extension/issues/1271 to track the issue on Windows. Thanks for your investigation.

**Comment by Zhirong2022** [comment]
> > test#.sas
> 
> Hey @Zhirong2022. Let's skip this one in this pull request. This is currently an issue that impacts sas content (in prod now) & sas server. It looks to be an issue with uri parsing. I'll create a follow-up PR to handle this case.

@scottdover ,  https://github.com/sassoftware/vscode-sas-extension/issues/1272 was raised to track the issue. Thanks.

**Comment by scottdover** [comment]
> @scottdover , please check the errors below with package generation. image

Hey @Zhirong2022. Sorry about that. I'm still attempting to resolve 

> Error 'The file type is unsupported.'is thrown if trying to convert a .sasnb file
> 1.Launch VSC and make sure the initial active profile is a non-Viya connection
> 2.Switch the profile to a Viya connection
> 3.Create a new .sasnb file on file server
> 4.Convert the .sasnb file to the .flw file

See the comment up above. Once that is marked as resolved, things should be ready to re-test

**Comment by Zhirong2022** [comment]
Drag/drop a file or files to a position where they have already there, it has no prompt error message in such scenario.  Perform the same against SAS Content, it has error validation. 
image

**Comment by scottdover** [comment]
**The following issues have been resolved in https://github.com/sassoftware/vscode-sas-extension/pull/1203/commits/7115d207195930a7c0bf54a46e44af313ec1c4f3**

> Better error validation for scenarios below: 1.Rename a file with the same name as an existed one Error message: Error running command SAS.server.renameResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.renameResource.
> 
> 2.Rename a file folder with the same name as an existed one Error message:Error running command SAS.server.renameResource: Request failed with status code 409. This is likely caused by the extension that contributes SAS.server.renameResource.

> It failed to save a file to the sever 1.Create a file folder named **NewFolder** 2.Create a file named **NewFile.sas** under **NewFolder** 3.Type some code in the **NewFile.sas** file 4.Delete **NewFolder** and **NewFild.sas** was kept unsaved 5.Click SaveAs.. on the prompt message 6.Type the correct path which is authorized image

**NOTE: Now open files _under_ a renamed folder will  automatically be closed and re-opened when the folder is renamed**

> Error is thrown if trying to convert the .sasnb file for the second time 
> 1.Select a .sasnb file and convert it to a .flw file with the default given name, it works well
> 2.Switch the profile to a new one (any connection type)
> 3.Switch the profile back
> 4.Delete the newly created .flw file
> 5.Convert the same .sasnb the second time without change the .flw name

> Error 'The file type is unsupported.'is thrown if trying to convert a .sasnb file 
> 1.Launch VSC and make sure the initial active profile is a non-Viya connection 
> 2.Switch the profile to a Viya connection 
> 3.Create a new .sasnb file on file server 
> 4.Convert the .sasnb file to the .flw file

> Drag/drop a file or files to a position where they have already been there, it has no prompt error message in such scenario. Perform the same against SAS Content, it has error validation. image

**This issue remains unresolved:**

> Create a new **.txt** file with name **!@$%^&*(**.txt, it has error if trying to download it through ContextMenu.
> Error message:
> Error running command SAS.server.downloadResource: Error: ENOENT: no such file or directory, open 'c:\\!@$: > %^&*(.txt'. This is likely caused by the extension that contributes SAS.server.downloadResource.

Can you try this with SAS content? Can you download via that? I'm guessing you can't. Fwiw, it isn't an issue on Mac, but is an issue on Windows. If it's not an issue specific to this story, I'll look at in a future changeset. But, Windows may be the limiting factor here.

**Comment by Zhirong2022** [comment]
Error happened if trying to rename a file.
image

**Comment by Zhirong2022** [comment]
Rename a file folder with the same name as an existed one.
image

**Comment by scottdover** [comment]
@Zhirong2022 The rename issues should now be resolved

**Comment by Zhirong2022** [comment]
New a file and keep it open, rename the filename
image

**Comment by scottdover** [comment]
> New a file and keep it open, rename the filename image

Resolved in https://github.com/sassoftware/vscode-sas-extension/pull/1203/commits/f93ced33ad1a6fa5dc2403078929d5cda540576a

**Comment by Zhirong2022** [comment]
It failed to drag and drop a file to a target file folder named with **NewFolder;** No such issue with SAS Content.
1. Create a new folder named with **NewFolder;**
2. Drag and drop a file to **NewFolder;**

Error on console for reference:
ERR Request failed with status code 404: AxiosError: Request failed with status code 404

**Comment by scottdover** [comment]
> It failed to drag and drop a file to a target file folder named with **NewFolder;** No such issue with SAS Content.
> 
>     1. Create a new folder named with **NewFolder;**
> 
>     2. Drag and drop a file to **NewFolder;**
> 
> 
> Error on console for reference: ERR Request failed with status code 404: AxiosError: Request failed with status code 404

This has been resolved in https://github.com/sassoftware/vscode-sas-extension/pull/1203/commits/d7e2e1bdf646e09bbcec0eef84820528ca4b7314. 

Imo, this seems like a pretty uncommon use-case (along with some of the other naming issues like having a file named `NewFile+!@$%^&*.txt`). Unless there is opposition from @clangsmith, I propose we aggregate any further issues like this into a separate bug where this PR can move along. I don't want to hold this release up for use cases that most of our users won't encounter.

@scnwwu @smorrisj Thoughts?

**Comment by Zhirong2022** [comment]
All issues have been handled appropriately.  Some enhancement like multi-deletion will be tracked in a new issue after the PR gets pushed.

**Review by Zhirong2022 — APPROVED** [review]

**Pull request merged**